src/python/tensor.py

```diff
@@ -0,0 +1,292 @@
+
+    '''
+    if arg == None:
+      # call constructor of singa::Tensor
+      self.singa_tensor = singa.Tensor()
+      return
+    elif isinstance(arg, Tensor):
+      # call copy constructor of singa::Tensor
+      self.singa_tensor = singa.Tensor(arg.singa_tensor)
+      self.tuple_shape = arg.tuple_shape
+      self.device = arg.device
+      self.dtype = arg.dtype
+      return
+    else:
+      # otherwise, shape is required
+      assert not type(arg) == int, 'shape should be tuple'
+      shape = arg
+
+    # call overloaded constructor of singa::Tensor
+    vs = _tuple_to_vector(shape)
+    if device == None:
+      self.singa_tensor = singa.Tensor(vs, dtype)
+    else:
+      self.singa_tensor = singa.Tensor(vs, device, dtype)
+
+    self.tuple_shape = shape
+    self.device = device
+    self.dtype = dtype
+
+  def data(self):
+    #TODO(chonho) - need to think more efficient way to convert???
+    idx = self.singa_tensor.data_type()
+    if idx == kFloat32:
+      data_array = singa.floatArray_frompointer(self.singa_tensor.floatData())
+      dt = np.float32
+    elif idx == kFloat16:
+      print 'not implemented yet'
+      return
+      #data_array = singa.floatArray_frompointer(self.singa_tensor.floatData())
+      #dt = np.float16
+    elif idx == kInt:
+      data_array = singa.intArray_frompointer(self.singa_tensor.intData())
+      dt = np.int32
+    elif idx == kChar:
+      data_array = singa.charArray_frompointer(self.singa_tensor.charData())
+      dt = np.int8
+    elif idx == kDouble:
+      data_array = singa.doubleArray_frompointer(self.singa_tensor.doubleData())
+      dt = np.float64
+
+    data = [data_array[i] for i in range(self.singa_tensor.Size())]
+    data = np.array(data, dtype=dt).reshape(self.tuple_shape)
+    return data
+
+  def data_type(self):
+    return self.singa_tensor.data_type()
+
+  def shape(self, axis=None):
+    if axis==None:
+       return self.singa_tensor.shape()
+    else:
+       return self.singa_tensor.shape(axis)
+
+  def ndim(self):
+    return self.singa_tensor.nDim()
+
+  def is_transpose(self):
+    return self.singa_tensor.transpose()
+
+  def size(self):
+    return self.singa_tensor.Size()
+
+  def memsize(self):
+    return self.singa_tensor.MemSize()
+
+  def reshape(self, shape):
+    assert product(self.tuple_shape) == product(shape), \
+           'product of shape should be equal'
+    self.tuple_shape = shape
+    self.singa_tensor.Reshape(_tuple_to_vector(shape))
+
+  def reset_like(self, t):
+    self.singa_tensor.ResetLike(t.singa_tensor)
+
+  def as_type(self, dtype):
+    self.singa_tensor.AsType(dtype)
+
+  def to_device(self, device):
+    self.singa_tensor.ToDevice(device)
+
+  def to_host(self):
+    self.singa_tensor.ToHost()
+
+  def set_value(self, x):
+    self.singa_tensor.SetValue(x)
+
+  def copy_data(self, t):
+    self.singa_tensor.CopyData(t.singa_tensor)
+
+  def clone(self):
+    new_t = Tensor(self.tuple_shape, self.device, self.dtype)
```

Comment:
For efficiency, just create an dummy tensor without passing shape, device or dtype.
after call new_t.singa_tensor=..clone..
set shape, device, and dtyep of new_t.